src/transformers/models/whisper/modeling_flax_whisper.py

```diff
@@ -0,0 +1,1500 @@
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.return_dict
+
+        # prepare decoder inputs
+        if decoder_attention_mask is None:
+            decoder_attention_mask = jnp.ones_like(decoder_input_ids)
+        if decoder_position_ids is None:
+            batch_size, sequence_length = decoder_input_ids.shape
+            decoder_position_ids = jnp.broadcast_to(
+                jnp.arange(sequence_length)[None, :], (batch_size, sequence_length)
+            )
+
+        # Handle any PRNG if needed
+        rngs = {"dropout": dropout_rng} if dropout_rng is not None else {}
+
+        return self.module.apply(
+            {"params": params or self.params},
+            input_features=input_features,
+            decoder_input_ids=jnp.array(decoder_input_ids, dtype="i4"),
+            decoder_attention_mask=jnp.array(decoder_attention_mask, dtype="i4"),
+            decoder_position_ids=jnp.array(decoder_position_ids, dtype="i4"),
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+            deterministic=not train,
+            rngs=rngs,
+        )
+
+    def scan(self, params: Optional[FrozenDict] = None):
+        self._module = self.module_class(config=self.config, dtype=self.dtype, use_scan=True)
+        init_fn = partial(self.init_weights, input_shape=self.input_shape)
+        params_shape_tree = jax.eval_shape(init_fn, self.key)
+
+        # get the shape of the parameters
+        self._params_shape_tree = params_shape_tree
+
+        # save required_params as set
+        self._required_params = set(flatten_dict(unfreeze(params_shape_tree)).keys())
+
+        # initialize the parameters
+
+        if params is not None:
+            params = convert_unroll_to_scan(self, params)
+            return params
+        elif self._is_initialized:
+            self.params = convert_unroll_to_scan(self, self.params)
+
+    def scan_disable(self, params: Optional[FrozenDict] = None):
+        self._module = self.module_class(
+            config=self.config,
+            dtype=self.dtype,
+            use_scan=False,
+        )
+        init_fn = partial(self.init_weights, input_shape=self.input_shape)
+        params_shape_tree = jax.eval_shape(init_fn, self.key)
+
+        # get the shape of the parameters
+        self._params_shape_tree = params_shape_tree
+
+        # save required_params as set
+        self._required_params = set(flatten_dict(unfreeze(params_shape_tree)).keys())
+
+        # initialize the parameters
+        if params is not None:
+            params = convert_scan_to_unroll(self, params)
+            return params
+        elif self._is_initialized:
+            self.params = convert_scan_to_unroll(self, self.params)
```

Comment:
Same comment for `scan`, would like to have @sgugger 's opinion but I don't think we should add such functionalities

Reply:
(copied from https://github.com/huggingface/transformers/pull/18341, see comment below with regards to scan https://github.com/huggingface/transformers/pull/20479#issuecomment-1335538699)